### ISSUE TITLE
"click to authenticate" button not clickable

### ISSUE THREAD

**Issue by Braamling** [springfox/springfox#870]
I'm trying to configure OAuth with springfox (followup on #834). The button appears but is not clickable.

There is a small error icon next to the button: 
```html
<div class="auth">
      <span class="api-ic ic-error"></span>
         <div id="api_information_panel" style="top: 526px; left: 776px; display: none;">
         <div title="read and write">write</div>
     </div>
</div>
```

There is no error message, and the information panel appears somewhere else on the page. 

```java
public Docket swaggerSpringMvcPlugin()
{
     return new Docket( DocumentationType.SWAGGER_2 )
        .groupName( "test" )
        .select()
            .apis( not( withMethodAnnotation( PrivateAPI.class )))
            .paths( regex( ".*/rest/.*" ) )
            .build().apiInfo( new ApiInfoBuilder()
                .title( "test" )
                .description( "test" )
                .version( "0.1" )
                .build())
		.securitySchemes(securitySchemes())
		.securityContexts(securityContext()); 

	private class PasswordTokenRequestEndpoint extends TokenRequestEndpoint {

		private final String _username;
		private final String _password;

		public PasswordTokenRequestEndpoint(String url, String clientIdName,
				String clientSecretName, String username, String password) {

			super(url, clientIdName, clientSecretName);
			this._username = username;
			this._password = password;
		}

		@SuppressWarnings("unused")
		public String getUsername() {
			return _username;
		}

		@SuppressWarnings("unused")
		public String getPassword() {
			return _password;
		}
	}

	/**
	 * Class used for Oauth2 password grant type
	 */
	private class OAuth2PasswordCredentialsGrantType extends GrantType {

		@SuppressWarnings("unused")
		private final PasswordTokenRequestEndpoint _tokenRequestEndpoint;

		public OAuth2PasswordCredentialsGrantType(
				PasswordTokenRequestEndpoint tokenRequestEndpoint) {
			super("password");
			this._tokenRequestEndpoint = tokenRequestEndpoint;
		}
	}
	
	private List<SecurityContext> securityContext() {
		return Arrays.asList(SecurityContext.builder()
		.securityReferences(SecurityReferences())
		.build());
	}
	
	private List<SecurityScheme> securitySchemes() {
		ArrayList<SecurityScheme> authorizationTypes = new ArrayList<>();

		/* not sure why scopes are used for both authorization and authorization
		 * type
		 * API is not using the scopes for now
		 */
		List<AuthorizationScope> authorizationScopeList = new ArrayList<>();
		authorizationScopeList.add(new AuthorizationScope("read", "read only"));
		authorizationScopeList.add(new AuthorizationScope("write", "read and write"));

		List<GrantType> grantTypes = new ArrayList<>();

		TokenRequestEndpoint tokenRequestEndpoint = new TokenRequestEndpoint("/oauth/token", "client_Id ", "client_secret" );
		TokenEndpoint tokenEndpoint = new TokenEndpoint("/oauth/token", "theToken");

		PasswordTokenRequestEndpoint passwordTokenRequestEndpoint = new PasswordTokenRequestEndpoint(
				"/oauth/token", "client_Id ", "client_secret", "user", "password");
		grantTypes.add(new OAuth2PasswordCredentialsGrantType(
				passwordTokenRequestEndpoint));
		
		grantTypes.add(new AuthorizationCodeGrant(tokenRequestEndpoint, tokenEndpoint));
		
		/* OAuth authorization type with client credentials and password grant
		 * types.
		 */
		authorizationTypes.add(new OAuth("oauth2", authorizationScopeList,
				grantTypes));

		 authorizationTypes.add(oAuth);
		
		return authorizationTypes;
	}

	private List<SecurityReference> SecurityReferences() {

		List<SecurityReference> authorizations = new ArrayList<>();

		/* use same scopes as above */
		AuthorizationScope[] authorizationScopes = {
				new AuthorizationScope("read", "read only"),
				new AuthorizationScope("write", "read and write") };

		/* Currently we have 2 roles - user and client */
		authorizations.add(new SecurityReference("USER", authorizationScopes));
		authorizations
				.add(new SecurityReference("CLIENT", authorizationScopes));

		return authorizations;

	}
```

An example controller method we use is:
```
	@RequestMapping	(value="/apiCallExample", method = RequestMethod.GET)
	@ApiOperation	(value = "Get event", response = RestAPI.class,
    				authorizations = {
							@Authorization(value = "oauth2", scopes = {
    								@AuthorizationScope(scope = "write", description = "read and write")
    						})})
```

I am still unable to find out why this is failing, mainly because there are no error messages showing.  Am I missing something?

**Comment by dilipkrish** [comment]
What version are you using? Could you try with 2.1.2?

**Comment by Braamling** [comment]
I am using version 2.1.1, is 2.1.2 already available in maven?

**Comment by dilipkrish** [comment]
It should be by now

**Comment by Braamling** [comment]
Upgrading to 2.1.2 didn't solve the issue, is there something I could be doing wrong? Or has the issue not been solved correctly?

**Comment by yukinami** [comment]
I have the same problem now after I updated to v2.1.2. But it's ok when v2.0.2 except the issue I mentioned in #833. I looked into the newest code in swagger-ui.html.  Method `initializeSpringfox` which initializes the OAuth related function is invoked after document ready, when the `swagger-ui-container` hasn't  been loaded yet.  Should not it invoked after `swagger-ui-container` loaded completely ?

**Comment by vsafin-copart** [comment]
This is really a bug: 

Ran into the same problem and I believe the issue is in swagger-ui.html

This function runs first to setup all authentication stuff:

      function initializeSpringfox() {
        var security = {};
        window.springfox.securityConfig(function(data) {
          security = data;
          if (security.apiKey) {
            $('#input_apiKey').val(security.apiKey);
            addApiKeyAuthorization();
          }
          if (typeof initOAuth == "function") {
            if (security.clientId && security.appName && security.realm) {
              initOAuth(security);
            }
          }
        });
Inside initOAuth you setting that button onClick event:

$('.api-ic').click(function(s) {
if($(s.target).hasClass('ic-off'))
handleLogin();
else {
handleLogout();
}
false;
});

Unfortunately - at that time NOTHING is loaded in

yet.
initializeSpringfox should be called in OnComplete handler here:

window.swaggerUi = new SwaggerUi({
dom_id: "swagger-ui-container",
supportedSubmitMethods: ['get', 'post', 'put', 'delete', 'patch'],
onComplete: function(swaggerApi, swaggerUi) {

          $('pre code').each(function(i, e) {
            hljs.highlightBlock(e)
          });

        },
        onFailure: function(data) {
          log("Unable to Load SwaggerUI");
        },
        docExpansion: "none",
        apisSorter: "alpha"
      });

**Comment by vsafin-copart** [comment]
And one more thing that needs to be fixed: Authentication server redirects back to o2c.html and it should redirect to webjars/springfox-swagger-ui/o2c.html... need to adjust

**Comment by ejain** [comment]
btw how do you get username and password form fields to show up? Right now, all I see in the UI is a single "api_key" field...

**Comment by dilipkrish** [comment]
@ejain Its supposed to show you an oauth dialog? Does that not work?

**Comment by ejain** [comment]
@dilipkrish there's no oauth dialog, just an "api_key" field and an "explore" button.

**Comment by dilipkrish** [comment]
If you've set up your security references and security context correctly, it should show you something like this
![Oauth button](https://cloud.githubusercontent.com/assets/249413/2771937/ea7c1920-ca73-11e3-835e-3d212430f4b6.png)

**Comment by ejain** [comment]
The toggle is there, too, but doesn't do anything (or generate any js errors)...

**Comment by dhoffi** [comment]
same problem on my code. Button is there, but is disabled. Any news on this?

**Comment by dilipkrish** [comment]
I'll reopen this and take a look. Would any of you guys be able to see what the error is in the javascript console? Or is it just that there is no error and nothing happens?

**Comment by kjohns3** [comment]
I ran into a similar problem.  When I follow the code in swagger-oauth.js the clientSecret is undefined. I found this configuration is pulled from /configuration/security endpoint, which is backed by the SecurityConfiguration class.  This class doesn't have a property for clientSecret, so it fails at the error check.  The logic attempts to call log() to send a "missing client secret" message to the console, but there isn't a log function so it fails without warning.

**Comment by alighm** [comment]
So I see the error as well. Basically, I see the "click to authenticate" button there but when clicking on it nothing happens. I went ahead and executed the following after the swagger-ui.html page render is complete:

 $('.api-ic').click(function(s) {
    if($(s.target).hasClass('ic-off'))
      handleLogin();
    else {
      handleLogout();
    }
    false;
  });

Base on this post: https://github.com/springfox/springfox/issues/767

Now after executing the above, when I click on the toggle I get a js error saying: 

Uncaught TypeError: Cannot read property 'show' of undefined on swagger-oauth.js line 163.

Basically popupMask is undefined, and popupMask should be $('#api-common-mask') I assume.

I hope this is useful.

**Comment by abelyaevskiy** [comment]
Hello!

@alighm, did you find workaround to enble the button?

@dilipkrish, do you plan to fix it?

**Comment by n0mer** [comment]
@abelyaevskiy  +1

**Comment by jbspeakr** [comment]
+1

**Comment by dilipkrish** [comment]
Verified this has been fixed in 2.3.0 (2.2.3-SNAPSHOT)

In addition to defining the security context and security schemes, it requires that you need to create a security configuration bean for this to work. Something like 

```java
  @Bean
  public SecurityConfiguration securityInfo() {
    return new SecurityConfiguration("abc", "123", "pets", "petstore", "123", ",");
  }
```

**Comment by n0mer** [comment]
@dilipkrish, can you please briefly explain the meaning of parameters?

Is it somehow related to "resource_ids, scope, authorized_grant_types, authorities" ?

String clientId - that's okay
String clientSecret - this is empty, in my case - i'm using "user_member" with login "roy" and password "spring"
String realm - what is it?
String appName - what is it? is it "resource_id"?
String apiKey - i'm not using apiKey
String scopeSeparator - this supposed to be always ','

**Comment by n0mer** [comment]
and looks like the snapshot http://oss.jfrog.org/artifactory/oss-snapshot-local/io/springfox/springfox-swagger-common/2.2.3-SNAPSHOT/springfox-swagger-common-2.2.3-20151205.005057-38.jar does not contain constructor for SecurityConfiguration that takes 6 params, only for 4

**Comment by dilipkrish** [comment]
@n0mer Just released 2.3.0. It should have the parameters available.

To be really honest I'm not much of an oauth expert. This is just a representation of all the configurable values that swagger-ui demands. Its a grab bag of configuration values and not all are related to each other. With that said I'll try to answer some of your questions

> String clientId - that's okay
> String clientSecret - this is empty, in my case - i'm using "user_member" with login "roy" and password "spring"
> String realm - what is it?

These three are for client credentials, tho' I don't know what the values should be for realm. 

> String appName - what is it? is it "resource_id"?

I believe that is what this is. 

> String scopeSeparator - this supposed to be always ','

Yes, This is independent of the client credentials grant. Just a way to configure a seperator for the demands. It is defaulted to ',' like you suggested.

**Comment by ralfstuckert** [comment]
Did anybody actually manage to get this to work? I'm using version 2.4.0 and the "click to authenticate" button does not respond :-\

Maybe someone could provide a working example, that would be quite helpful

**Comment by dilipkrish** [comment]
@ralfstuckert try 2.5.0... if it doesn't work please create a new issue

**Comment by ralfstuckert** [comment]
Ok, button responds now. But the login URL is not set up correctly. It opens the URL
http://**localhost:9001/null**&redirect_uri=http%3A%2F%2Flocalhost...

The setup of the login URL is as follows:

@Bean
public Docket api() {
...
     .securitySchemes(Collections.singletonList(securitySchema()))
...
}

private OAuth securitySchema() {
    final AuthorizationScope authorizationScope =
        new AuthorizationScope(authorizationScopeGlobal, authorizationScopeGlobalDesc);
    final LoginEndpoint loginEndpoint = new LoginEndpoint(loginUrl);
    final GrantType grantType = new ImplicitGrant(loginEndpoint, "access_token");
    return new OAuth("oauth2", Collections.singletonList(authorizationScope ), 
        Collections.singletonList(grantType));
}

Where's my fault?

**Comment by dilipkrish** [comment]
what does your login url look like?

**Comment by ralfstuckert** [comment]
http://127.0.0.1:9008/dialog.html

**Comment by toddcscar** [comment]
Can the dialog be used for Authorization headers (e.g. Bearer) or is it strictly meant for Basic Auth header?  i am also trying to enable my app to display the authenticate link but it isn't working.

**Comment by vroldanbet** [comment]
Still the same issue in 2.5.0

**Comment by tianshaofandine** [comment]
@ralfstuckert How did you get yours working?

**Comment by ralfstuckert** [comment]
I'm using springfox 2.5.0 with implicit flow and a login dialog. My swagger config looks like this: 
```
@Configuration
@EnableSwagger2
public class SwaggerConfig extends WebMvcConfigurerAdapter {

	@Value("${dvm.login.url}")
	private String loginUrl;
	@Value("${oauth2.swaggerClientId}")
	private String swaggerClientId;
	@Value("${oauth2.swaggerClientSecret}")
	private String swaggerClientSecret;

	@Value("${info.build.description}")
	private String infoBuildDescription;
	@Value("${info.build.artifact}")
	private String infoBuildArtifact;
	@Value("${info.build.name}")
	private String infoBuildName;
	@Value("${info.build.version}")
	private String infoBuildVersion;

	public static final String securitySchemaOAuth2 = "oauth2";
	public static final String authorizationScopeGlobal = "global";
	public static final String authorizationScopeGlobalDesc = "accessEverything";

	@Override
	public void addViewControllers(final ViewControllerRegistry registry) {
		registry.addViewController("/").setViewName("redirect:/swagger-ui.html");
	}


	@Bean
	public Docket api() {
		return new Docket(DocumentationType.SWAGGER_2). //
				directModelSubstitute(LocalDate.class, java.sql.Date.class). //
				useDefaultResponseMessages(false). //
				apiInfo(getApiInfo()). //
				select(). //
				apis(RequestHandlerSelectors.any()).paths(PathSelectors.ant("/api/**")). //
				build(). //
				securitySchemes(Collections.singletonList(securitySchema()))
				.securityContexts(Collections.singletonList(securityContext()));
	}

	private ApiInfo getApiInfo() {
		return new ApiInfo(infoBuildName, infoBuildDescription, infoBuildVersion, "",
				(Contact) null, "", "");
	}

	private OAuth securitySchema() {
		final LoginEndpoint loginEndpoint = new LoginEndpoint(loginUrl);
		final GrantType grantType = new ImplicitGrant(loginEndpoint, "access_token");
		final List<AuthorizationScope> scopes = new ArrayList<>();
		scopes.add(new AuthorizationScope("sampleScope, "there must be at least one scope here"));
		return new OAuth(securitySchemaOAuth2, scopes, Collections.singletonList(grantType));
	}


	private SecurityContext securityContext() {
		return SecurityContext.builder().securityReferences(defaultAuth())
				.forPaths(PathSelectors.ant("/api/**")).build();
	}

	private List<SecurityReference> defaultAuth() {
		final AuthorizationScope authorizationScope =
				new AuthorizationScope(authorizationScopeGlobal, authorizationScopeGlobalDesc);
		final AuthorizationScope[] authorizationScopes = new AuthorizationScope[1];
		authorizationScopes[0] = authorizationScope;
		return Collections
				.singletonList(new SecurityReference(securitySchemaOAuth2, authorizationScopes));
	}

	@Bean
	public SecurityConfiguration securityInfo() {
		return new SecurityConfiguration(swaggerClientId, swaggerClientSecret, "realm", infoBuildName, "", ApiKeyVehicle.HEADER, "api_key", ",");
	}
}
```

There is still a [problem in swagger-ui](https://github.com/swagger-api/swagger-ui/issues/2296), that's why I'm using a patched version of it.

Hope this help's
Ralf

**Comment by huiwang** [comment]
@ralfstuckert can you explain how you implemented the dialog page ?

**Comment by ralfstuckert** [comment]
@huiwang it is a simple form that uses the entered credentials to create an access token via an ajax call. Then the window.location is replaced with the passed redirect URL where the access_token is appended as a query parameter.

```
window.location.replace(redirectUri + "?access_token=" + data.access_token);
```

**Comment by huiwang** [comment]
@ralfstuckert thx for your fast response, would you mind sending me your login html ?

**Comment by ralfstuckert** [comment]
Sorry, since this is a customer project I cannot hand out complete code files

**Comment by huiwang** [comment]
@ralfstuckert no problem, i'll try to figure it out :) thx a lot for your help

**Comment by thirupathi-redpinesignals** [comment]
is there any swagger documentation to build api authentication using oauth2 and springboot similar to http://petstore.swagger.io

**Comment by dilipkrish** [comment]
@thirupathi-redpinesignals Have you seen [the docs](http://springfox.github.io/springfox/docs/current/#springfox-swagger2-with-spring-mvc-and-spring-boot)?

**Comment by bojan007** [comment]
Hi, I'm using springfox 2.5.0 and I'm trying to make the authorization with oauth2 work, but the authenticate button isn't clickable. Is there still an issue or is my configuration incorrect? How can I check if my configuration is causing the problem?

**Comment by dilipkrish** [comment]
@bojan007 Not really sure. There would be clues in your javascript console

**Comment by gazeciarz** [comment]
Thank You @ralfstuckert to better understand the topic. Works well with 2.6.0 and Your example. Bad thing is there is not much tutorials how to integrate OAuth2 with swagger...

**Comment by dilipkrish** [comment]
[Jhipster](https://jhipster.github.io) does a good job of crystalizing the best practices...  Try generating a sample and see how they do it

**Comment by blelem** [comment]
[JHipster](https://jhipster.github.io/) is indeed good example how OAuth2, swagger and springfox can play well together. That helped me a great deal. The key take away is that they do not use springfox security configuration at all. Before looking at jHipster, my mind was set that in order to use OAuth2 with swagger, one must configure securitySchemes, securityContexts and others OAuth related config as defined in [springfox's docs](http://springfox.github.io/springfox/docs/current/#springfox-swagger2-with-spring-mvc-and-spring-boot). And the discussion above somewhat reinforced that idea. But that is not the case, and I would actually strongly advocate to stay clear from springfox security config, if one can do without.
This is how JHipster are doing:
 1. They use the site's existing infrastructure for the login. They expect that the user will first log in to the site, as any user would normally do. When user has successfully logged in, they save the user's OAuth2 authentication token to the app's local storage, from their [client side javascript code](https://gist.github.com/blelem/dd9c9a25748ae1dcda11d942132f2595#file-auth-oauth2-service-js).
 ```javascript
 function authSucess (response) {
     var expiredAt = new Date();
     expiredAt.setSeconds(expiredAt.getSeconds() + response.expires_in);
     response.expires_at = expiredAt.getTime();
     $localStorage.authenticationToken = response;
     return response;
  }
```
 2. When serving the swagger url, they configure the SwaggerClient to use OAuth2 Bearer token, while injecting the saved OAuth token. The served swagger html page looks like [this](https://gist.github.com/blelem/dd9c9a25748ae1dcda11d942132f2595#file-index-html).
```javascript
 function addApiKeyAuthorization(){
      var authToken = JSON.parse(localStorage.getItem("jhi-authenticationToken")).access_token;
      var apiKeyAuth = new SwaggerClient.ApiKeyAuthorization("Authorization",
                            "Bearer " + authToken, "header");
      window.swaggerUi.api.clientAuthorizations.add("key", apiKeyAuth);
  }
```
Thus, no Springfox OAuth2 config, everything is configured on the client's javascript code.

**Comment by dilipkrish** [comment]
@blelem also note that the key thing is also they don't use the bundled Springfox swagger ui. They use their own custom swagger ui so none of the security customizations that the docket provides makes a difference. 

Springfox provides a very limited oauth support and it doesn't support all the oauth profiles.

**Comment by blelem** [comment]
@dilipkrish Mmm, I haven't noticed the custom swaggerui yet. Upon a very superficial check, they seem to pull the mainstream swagger ui using bower without patching it. But as I'm still integrating their solution into mine, I might have missed their customization.

**Comment by blelem** [comment]
@dilipkrish, it's definitely good to stress out that the swaggerui's index.html used by JHipster is not the same as provided by Springfox. I've updated my 'answer' above.
